File: javascript_implementation/movie_trends.js
const sequelize = require("./models/db");
const Movie = require("./models/movie");
const Theater = require("./models/theater");
const Sale = require("./models/sale");
const prompt = require("prompt-sync")();
const { QueryTypes } = require("sequelize");

async function create_session() {
  /*
  Sets up the in-memory database.

  Args:
    None
  Returns:
    Nones
*/
  await sequelize.sync({ force: true });
}

async function create_database() {
  /*
  Populates the in-memory database with sample data.

  Args:
    None
  Returns:
    Nones
*/
  const movie_1 = await sequelize.models.movies.create({
    name: "Movie 1",
  });
  const movie_2 = await Movie.create({
    name: "Movie 2",
  });
  const theater_1 = await Theater.create({
    name: "Theater 1",
  });
  const theater_2 = await Theater.create({
    name: "Theater 2",
  });
  const sale_1 = await Sale.create({
    date: new Date(2025, 1, 24, 19), // Adding 19 hours to the date so it doesn't change upon conversion to UTC by default
    movie_id: movie_1.id,
    theater_id: theater_1.id,
    tickets_sold: 100,
  });
  const sale_2 = await Sale.create({
    date: new Date(2025, 1, 24, 19),
    movie_id: movie_1.id,
    theater_id: theater_2.id,
    tickets_sold: 200,
  });
  const sale_3 = await Sale.create({
    date: new Date(2025, 1, 24, 19),
    movie_id: movie_2.id,
    theater_id: theater_1.id,
    tickets_sold: 150,
  });
  const sale_4 = await Sale.create({
    date: new Date(2025, 1, 24, 19),
    movie_id: movie_2.id,
    theater_id: theater_2.id,
    tickets_sold: 250,
  });
  const sale_5 = await Sale.create({
    date: new Date(2025, 1, 25, 19),
    movie_id: movie_1.id,
    theater_id: theater_1.id,
    tickets_sold: 500,
  });
  const sale_6 = await Sale.create({
    date: new Date(2025, 1, 25, 19),
    movie_id: movie_1.id,
    theater_id: theater_2.id,
    tickets_sold: 400,
  });
  const sale_7 = await Sale.create({
    date: new Date(2025, 1, 25, 19),
    movie_id: movie_2.id,
    theater_id: theater_1.id,
    tickets_sold: 350,
  });
  const sale_8 = await Sale.create({
    date: new Date(2025, 1, 25, 19),
    movie_id: movie_2.id,
    theater_id: theater_2.id,
    tickets_sold: 450,
  });
}

function get_user_input() {
  /*
    Prompts the user for a date for which to see the best selling theater. Continues asking until a valid date is provided.

    Args:
        None
    Returns:
        Date: The valid date entered by the user.
 */
  let validDate;
  console.log(
    "Enter the date (MM/DD/YYYY) for which you want to see the best selling theater:"
  );
  let date = prompt("");
  while (true) {
    validDate = new Date(
      parseInt(date.split("/")[2]),
      parseInt(date.split("/")[0] - 1), // -1 because months are 0-indexed
      parseInt(date.split("/")[1])
    );
    // Taken from https://stackoverflow.com/questions/1353684/detecting-an-invalid-date-date-instance-in-javascript
    if (
      validDate.getFullYear() != date.split("/")[2] ||
      validDate.getMonth() != date.split("/")[0] - 1 ||
      validDate.getDate() != date.split("/")[1]
    ) {
      console.log(
        "Invalid date format. Please enter the date in MM/DD/YYYY format:"
      );
      date = prompt();
    } else break;
  }
  return validDate;
}

function convert_date_to_sql_format(date) {
  /* 
  Helper function to convert date to YYYY-MM-DD format 
  
  Args:
    date (Date): The date to convert.
  Returns:
    string: The date in YYYY-MM-DD format.
  */
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(
    2,
    "0"
  )}-${String(date.getDate()).padStart(2, "0")}`;
}

async function get_best_selling_theater(date) {
  /* 
    Retrieves and the best selling theater(s) for a given date.

    Args:
        date (Date): The date for which to find the best selling theater.

    Returns:
        None
    */
  const results = await sequelize.query(
    `SELECT b.id, b.name, SUM(a.tickets_sold) as tickets_sold FROM 'sales' as a JOIN 'theaters' as b on a.theater_id = b.id WHERE a.date = '${convert_date_to_sql_format(
      date
    )}' GROUP BY b.id  HAVING SUM(a.tickets_sold) = (SELECT SUM(tickets_sold) FROM 'sales' WHERE date = '${convert_date_to_sql_format(
      date
    )}' GROUP BY theater_id ORDER BY 1 DESC LIMIT 1)`,
    {
      type: QueryTypes.SELECT,
    }
  );
  print_results(results, date);
}

function print_results(results, date) {
  /*
    Neatly prints the best selling theater(s) for a given date.

    Args:
        results (List[(number: theater ID, string: theater name, number: tickets sold)]): 
                                                The results of the query of the theater(s) with the most tickets sold on the given date.
        date (Date): The date for which to find the best selling theater.

    Returns:
        None
    */
  if (results.length) {
    // Handles the case where there are multiple theaters with the same max sales
    console.log(
      `The best selling theater${
        results.length > 1 ? "s" : ""
      } on ${convert_date_to_sql_format(date)} ${
        results.length > 1 ? "are" : "is"
      }:`
    );
    results.forEach((result) => {
      console.log(`${result.name} with ${result.tickets_sold} tickets sold.`);
    });
  } else {
    console.log(
      `No sales data available for ${convert_date_to_sql_format(date)}`
    );
  }
}

async function main() {
  const date = get_user_input();
  await create_session();
  await create_database();
  await get_best_selling_theater(date);
}

main();

File: python_implementation/movie_trends.py
from sqlalchemy import create_engine, func
from sqlalchemy.orm import sessionmaker, Session
from models import Base
from models.sale import Sale
from models.movie import Movie
from models.theater import Theater
from datetime import datetime
from typing import List, Tuple


def create_session() -> Session:
    """
    Sets up the in-memory database.

    Returns:
        Session: A session object for interacting with the database.
    """
    engine = create_engine("sqlite:///:memory:")
    Base.metadata.create_all(engine)
    SessionLocal = sessionmaker(bind=engine)
    return SessionLocal()


def create_database(session: Session) -> None:
    """
    Populates the in-memory database with sample data.

    Args:
        session (Session): The SQLAlchemy session to use for database operations.

    Returns:
        None
    """
    movie_1 = Movie(name="Movie 1")
    movie_2 = Movie(name="Movie 2")
    theater_1 = Theater(name="Theater 1")
    theater_2 = Theater(name="Theater 2")
    session.add_all([movie_1, movie_2, theater_1, theater_2])
    session.flush()
    sale_1 = Sale(
        date=datetime(2025, 2, 24),
        movie=movie_1,
        theater=theater_1,
        tickets_sold=100,
    )
    sale_2 = Sale(
        date=datetime(2025, 2, 24),
        movie=movie_1,
        theater=theater_2,
        tickets_sold=200,
    )
    sale_3 = Sale(
        date=datetime(2025, 2, 24),
        movie=movie_2,
        theater=theater_1,
        tickets_sold=150,
    )
    sale_4 = Sale(
        date=datetime(2025, 2, 24),
        movie=movie_2,
        theater=theater_2,
        tickets_sold=250,
    )
    sale_5 = Sale(
        date=datetime(2025, 2, 25),
        movie=movie_1,
        theater=theater_1,
        tickets_sold=500,
    )
    sale_6 = Sale(
        date=datetime(2025, 2, 25),
        movie=movie_1,
        theater=theater_2,
        tickets_sold=400,
    )
    sale_7 = Sale(
        date=datetime(2025, 2, 25),
        movie=movie_2,
        theater=theater_1,
        tickets_sold=350,
    )
    sale_8 = Sale(
        date=datetime(2025, 2, 25),
        movie=movie_2,
        theater=theater_2,
        tickets_sold=450,
    )
    session.add_all([sale_1, sale_2, sale_3, sale_4, sale_5, sale_6, sale_7, sale_8])
    session.commit()


def get_user_input() -> datetime:
    """
    Prompts the user for a date for which to see the best selling theater. Continues asking until a valid date is provided.

    Returns:
        datetime: The valid date entered by the user.
    """
    valid_date = None
    print(
        "Enter the date (MM/DD/YYYY) for which you want to see the best selling theater:"
    )
    date = input()
    while not valid_date:
        try:
            valid_date = datetime.strptime(date, "%m/%d/%Y")
        except ValueError:
            print("Invalid date format. Please enter the date in MM/DD/YYYY format:")
            date = input()
    return valid_date


def get_best_selling_theater_results(session: Session, date: datetime) -> None:
    """
    Retrieves and the best selling theater(s) for a given date.

    Args:
        session (Session): The SQLAlchemy session to use for database operations.
        date (datetime): The date for which to find the best selling theater.

    Returns:
        None
    """
    # Subquery to find the maximum tickets sold for each theater on the given date
    max_subquery = (
        session.query(func.sum(Sale.tickets_sold).label("total"))
        .filter(Sale.date == date)
        .group_by(Sale.theater_id)
        .subquery()
    )
    # Scalar subquery to find the maximum total tickets sold over all theaters on the given date
    max_total = session.query(func.max(max_subquery.c.total)).scalar_subquery()
    results = (
        session.query(Theater, func.sum(Sale.tickets_sold))
        .join(Sale, Sale.theater_id == Theater.id)
        .filter(Sale.date == date)
        .group_by(Sale.theater_id)
        .having(func.sum(Sale.tickets_sold) == max_total)
        .all()
    )
    print_results(results, date)


def print_results(results: List[Tuple[Theater, int]], date: datetime) -> None:
    """
    Neatly prints the best selling theater(s) for a given date.

    Args:
        results (List[Tuple[Theater, int]]): The results of the query of the theater(s) with the most tickets sold on the given date.
        date (datetime): The date for which to find the best selling theater.

    Returns:
        None
    """
    if results:
        # Handles the case where there are multiple theaters with the same max sales
        print(
            f"The best selling theater{"s" if len(results) > 1 else ""} on {date.strftime('%m/%d/%Y')} {"are" if len(results) > 1 else "is"}:"
        )
        for theater, tickets_sold in results:
            print(f"{theater.name} with {tickets_sold} tickets sold.")
    else:
        print(f"No sales data available for {date.strftime('%m/%d/%Y')}.")


if __name__ == "__main__":
    date: datetime = get_user_input()
    with create_session() as session:
        create_database(session)
        get_best_selling_theater_results(session, date)
